### PR TITLE
Add get_auth_method utility function

### PR DESCRIPTION
**Pull request by pickfire**

### CODE REVIEW

**Comment by Minoru** [comment]
Most of Travis builds failed, can you fix them please?

File: rust/libnewsboat-ffi/src/utils.rs [hunk 1]
```diff
@@ -228,6 +228,11 @@ pub extern "C" fn rs_get_random_value(rs_max: u32) -> u32 {
     abort_on_panic(|| utils::get_random_value(rs_max))
 }
 
+#[no_mangle]
+pub extern "C" fn rs_get_auth_method(method: &str) -> u64 {
```
Comment (Minoru) [inline rust/libnewsboat-ffi/src/utils.rs]: FFI functions only accept C types. In this case, `*const c_char` instead of `&str`. See `rs_quote_if_necessary` above for an example on how to convert the C type into a Rust string. A comment on `unwrap()` is mandatory, and should explain why that particular call won't panic under normal conditions.

File: rust/libnewsboat/src/utils.rs [hunk 2]
```diff
@@ -311,6 +312,26 @@ pub fn is_valid_podcast_type(mimetype: &str) -> bool {
...
+        "digest" => curl_sys::CURLAUTH_DIGEST,
+        "digest_ie" => curl_sys::CURLAUTH_DIGEST_IE,
+        "gssnegotiate" => curl_sys::CURLAUTH_GSSNEGOTIATE,
+        "ntlm" => curl_sys::CURLAUTH_NTLM,
+        "anysafe" => curl_sys::CURLAUTH_ANYSAFE,
+        "any" | "" => curl_sys::CURLAUTH_ANY,
+        _ => {
+            log!(
+                Level::UserError,
+                "utils::get_auth_method: you configured an invalid proxy type: {}",
```
Comment (Minoru) [inline rust/libnewsboat/src/utils.rs]: Please change the string to what it said in C++: "you configured an invalid proxy authentication method". "Proxy type" is not the same as "proxy authentication method".

File: rust/libnewsboat/src/utils.rs [hunk 3]
```diff
@@ -731,6 +752,14 @@ mod tests {
         }
     }
 
+    #[test]
+    fn t_get_auth_method() {
```
Comment (Minoru) [inline rust/libnewsboat/src/utils.rs]: Coming up with additional tests is fine, but please port the existing ones as well. Having identical tests on both ends serves a few purposes:
* if e.g. Rust suite passes but C++ one fails, we'll know there is a problem with FFI;
* eventually we will remove the FFI wrapper for this function and drop C++ tests for it, which means we need to have a comprehensive test suite in Rust already—otherwise we'll lose test coverage.

File: rust/libnewsboat-ffi/src/utils.rs (hunk 1 above)
Comment (pickfire) [reply rust/libnewsboat-ffi/src/utils.rs]: Thanks for reminding me that, I will send in the changes later. But which `unwrap()` do you mean?

File: rust/libnewsboat/src/utils.rs (hunk 2 above)
Comment (pickfire) [reply rust/libnewsboat/src/utils.rs]: Ops, I forgot to change that.

File: rust/libnewsboat/src/utils.rs (hunk 3 above)
Comment (pickfire) [reply rust/libnewsboat/src/utils.rs]: I forgot to check the C++ tests for this, maybe later on will drop the C++ tests.

File: rust/libnewsboat/src/utils.rs [hunk 4]
```diff
@@ -311,6 +312,26 @@ pub fn is_valid_podcast_type(mimetype: &str) -> bool {
...
+        "digest" => curl_sys::CURLAUTH_DIGEST,
+        "digest_ie" => curl_sys::CURLAUTH_DIGEST_IE,
+        "gssnegotiate" => curl_sys::CURLAUTH_GSSNEGOTIATE,
+        "ntlm" => curl_sys::CURLAUTH_NTLM,
+        "anysafe" => curl_sys::CURLAUTH_ANYSAFE,
+        "any" | "" => curl_sys::CURLAUTH_ANY,
+        _ => {
+            log!(
+                Level::UserError,
+                "utils::get_auth_method: you configured an invalid proxy authenticate method: {}",
```
Comment (Minoru) [inline rust/libnewsboat/src/utils.rs]: Sorry, this still says "authenticate" rather than "authentication". I feel like I'm nitpicking, but c'mon, we had a proper message, why not just copy it.

File: rust/libnewsboat-ffi/src/utils.rs (hunk 1 above)
Comment (Minoru) [reply rust/libnewsboat-ffi/src/utils.rs]: Oh, my bad, you won't need an `unwrap` here.

File: rust/libnewsboat/src/utils.rs (hunk 4 above)
Comment (pickfire) [reply rust/libnewsboat/src/utils.rs]: Sorry my bad, I rarely copy paste when the sentence is short.

**Comment by Minoru** [comment]
Looks good code-wise, but it doesn't build with Rust 1.25 because of libz-sys crate ([macOS build](https://travis-ci.org/newsboat/newsboat/jobs/531149719#L2809), [Linux build](https://travis-ci.org/newsboat/newsboat/jobs/531149724#L1036)), and [is not formatted properly](https://travis-ci.org/newsboat/newsboat/jobs/531149737#L475). Will you deal with these, or should I take over? I usually just `rustup toolchain override set 1.25` and pick the latest version of a crate that still builds.

**Comment by pickfire** [comment]
> Looks good code-wise, but it doesn't build with Rust 1.25 because of libz-sys crate ([macOS build](https://travis-ci.org/newsboat/newsboat/jobs/531149719#L2809), [Linux build](https://travis-ci.org/newsboat/newsboat/jobs/531149724#L1036)), and [is not formatted properly](https://travis-ci.org/newsboat/newsboat/jobs/531149737#L475). Will you deal with these, or should I take over? I usually just `rustup toolchain override set 1.25` and pick the latest version of a crate that still builds.

I have yet test with 1.25, but what I usually do is `echo 1.25.0 > rust-toolchain`, I only do `rustup override set 1.25.0` sometimes. Maybe let me try again.

**Comment by pickfire** [comment]
Maybe you can take this on, I tried on this but not sure how to specify the dependency of dependency.

**Comment by Minoru** [comment]
Please don't rebase onto other PRs—that pollutes this PR's history with unrelated commits. Only rebase on `master`. 

> not sure how to specify the dependency of dependency

As far as I know, this can't be done directly, i.e. one can't say "curl-sys should be built with libz-sys version 1.2.3". However, one can take advantage of the fact that only one version of each crate gets built. By adding a `libz-sys` dependency directly to `libnewsboat` crate, I was able to "downgrade" it to a version that's buildable with Rust 1.25. I had to downgrade curl-sys a bit, too, since versions after 0.4.5 require libz-sys 1.0.18+.

I think it's time for me to figure out a real policy about compiler versions we support. Reading up on that now. Will probably open an issue about that later.

This PR looks good to me, I'll merge it when the CI passes.

**Comment by pickfire** [comment]
Sorry, I rebased onto my own master, not the origin master.

**Comment by Minoru** [comment]
[Many—but not all—builds failed]https://travis-ci.org/newsboat/newsboat/builds/531939383) because linker couldn't resolve openssl references. Interestingly, we didn't observe that with [an earlier build](https://travis-ci.org/newsboat/newsboat/builds/531591960) that used the latest curl-sys crate.

I don't want to give up on Rust 1.25 just yet, so I'm trying another solution: link libnewsboat.a before libssl is linked. Let's just hope that curl-sys and our own config.sh pick the same SSL library.

**Comment by Minoru** [comment]
Well, that didn't work out, and I've ran out of screen time for today. Will carry on tomorrow.

**Comment by pickfire** [comment]
I saw that debian is using rust 1.24.1. Why not stick to edition 2015 without using all the new 2018 features instead of sticking to a rust version since you need an old compiler.

**Comment by Minoru** [comment]
I aimed for 1.24.1 at first, but IIRC `once_cell` made me bump it to 1.24.1. I didn't fully comprehend Cargo.lock and Cargo's build planning back then, otherwise I'd try and force the 1.24.1 support by using an earlier `once_cell` version. Too late for that now.

We don't use any Rust 2018 features in Newsboat, primarily because they won't compile with 1.25. So we do stick to 2015 edition already. The problem is with dependencies that don't, and also dependencies that require newer compilers. The only way for us to fight that is to use older crate versions that have lower compiler requirements. But as we see here, it doesn't always work. I'm still looking into what can be done about that.

**Comment by Minoru** [comment]
Looks like I fixed it! Apparently the debug build of `curl-sys` (or one of its dependencies, most probably `openssl-sys`) requires a symbol from OpenSSL's `libssl` library. We now link `libssl` after `libnewsboat`, which fixes the problem. OSX might be using different code, so it wasn't affected at all. This is all just speculation, of course—I didn't dig into the code to find the exact line that's using that symbol.

@pickfire, thank you for this PR! Wanna try rewriting another `utils` function? ;)

**Comment by pickfire** [comment]
@Minoru Not at the moment, I am writing a parser for yaml. https://github.com/dtolnay/request-for-implementation/issues/9
